pkg/sbom/diff_nodelist.go

```diff
@@ -0,0 +1,234 @@
+// that are different in nl2 from nl after stripping identifier Quilifiers.
+func (nl *NodeList) DiffWithStripQulifiers(nl2 *NodeList) NodeListDiff {
+	return NodeListDiff{
+		NodesDiff:         nl.diffNodes(nl),
+		EdgesDiff:         nl.diffEdges(nl2),
+		RootElmementsDiff: nl.diffRootElements(nl2),
+	}
+}
+
+// diffRootElements computes the differences in root elements between two NodeLists.
+func (nl *NodeList) diffRootElements(nl2 *NodeList) NodeListRootElementDiff {
+	diff := NodeListRootElementDiff{}
+
+	nlRoots := make(map[string]struct{}) // map to store root elements of nl
+	for _, root := range nl.RootElements {
+		nlRoots[root] = struct{}{}
+	}
+
+	nl2Roots := make(map[string]struct{}) // map to store root elements of nl2
+	for _, root := range nl2.RootElements {
+		nl2Roots[root] = struct{}{}
+	}
+
+	// Compare root elements and populate the diff
+	for root := range nlRoots {
+		if _, exists := nl2Roots[root]; !exists {
+			diff.Removed = append(diff.Removed, root)
+		}
+	}
+	for root := range nl2Roots {
+		if _, exists := nlRoots[root]; !exists {
+			diff.Added = append(diff.Added, root)
+		}
+	}
+
+	sort.Strings(diff.Added)
+	sort.Strings(diff.Removed)
+
+	return diff
+}
+
+func (nl *NodeList) diffNodes(nl2 *NodeList) NodeListDiffNodes {
+	diff := NodeListDiffNodes{}
+
+	nlNodes := nl.Nodes
+	nl2Nodes := nl2.Nodes
+
+	// Sort both node lists
+	sort.Slice(nlNodes, func(i, j int) bool {
+		return strings.Compare(nlNodes[i].Id, nlNodes[j].Id) <= 0
+	})
+	sort.Slice(nl2Nodes, func(i, j int) bool {
+		return strings.Compare(nl2Nodes[i].Id, nl2Nodes[j].Id) <= 0
+	})
+
+	index1, index2 := 0, 0
+
+	// Iterate through both sorted node lists
+	for index1 < len(nlNodes) || index2 < len(nl2Nodes) {
+		switch {
+		case index1 < len(nlNodes) && index2 < len(nl2Nodes):
+			n, n2 := nlNodes[index1], nl2Nodes[index2]
+			switch strings.Compare(n.Id, n2.Id) { // Use ID to decide if to compare
+			case 0: // Nodes are equal
+				index1++
+				index2++
+				nodeDiff := n.Diff(n2)
+				if nodeDiff != nil {
+					diff.NodeDiff = append(diff.NodeDiff, nodeDiff)
+				}
+			case -1: // Node1 is less than Node2
+				diff.Removed = append(diff.Removed, n)
+				index1++
+			case 1: // Node1 is greater than Node2
+				diff.Added = append(diff.Added, n2)
+				index2++
+			}
+		case index1 < len(nlNodes):
+			diff.Removed = append(diff.Removed, nlNodes[index1])
+			index1++
+		default:
+			diff.Added = append(diff.Added, nl2Nodes[index2])
+			index2++
+		}
+	}
+
+	sort.Slice(diff.Added, func(i, j int) bool {
+		return strings.Compare(diff.Added[i].Id, diff.Added[j].Id) <= 0
+	})
+	sort.Slice(diff.Removed, func(i, j int) bool {
+		return strings.Compare(diff.Removed[i].Id, diff.Removed[j].Id) <= 0
+	})
+
+	sort.Slice(diff.NodeDiff, func(i, j int) bool {
+		return strings.Compare(diff.NodeDiff[i].Added.flatString(), diff.NodeDiff[j].Added.flatString()) <= 0
+	})
+
+	return diff
+}
+
+func (nl *NodeList) diffEdges(nl2 *NodeList) NodeListDiffEdges {
```

Comment:
This function is a nice start but it needs a little bit more work.

To generate the edge diff we need to compare and extract individual graph connections sharing the origin and type. For example, if I diff these two:

```go
Edge{
  "from": "node1"
  "type": "contains"
  "to": []string{"nodeA", "nodeB"}
 } 
Edge{
  "from": "node1"
  "type": "contains"
  "to": []string{"nodeA"}
 }
```
I should get the following diff:
```go
EdgeSetDiff{
  Removed: []Edge{
    Edge {
      "from": "node1"
      "type": "contains"
      "to": []string{"nodeB"}
    },
 }
```

 Also, we may need to work on an edge consolidation function, see this [comment about consolidation](https://github.com/protobom/storage/pull/4#issuecomment-2148705949) on the storage repo.